### PR TITLE
Move initTagField function from core.js to a initTagField.ts & add un…

### PR DESCRIPTION
**Pull request by xomcodes**
When complete closes at #9496
- [x] Move the function to the the new TypeScript file and set up correct types - https://www.typescriptlang.org/
- [x] Add a [JSDOC](https://jsdoc.app/) description to the function
- [x] Add a basic set unit test (just one is fine) for this function in a new client/src/includes/initTagField.test.js file using [Jest](https://jestjs.io/)
- [x] Ensure the util is imported in core.js correctly and the window variables window.initTagField works, take as screenshot of a tag field working after your change for the PR

…it test

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
-   [x] Run `make lint` from the Wagtail root.
-   [x] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
-   [x] **Please list the exact browser and operating system versions you tested**: Chrome version

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: client/src/includes/initTagField.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+/** 
+ *This is a function initTagField
+ *@function
+ *@param {number} id - The object id.
+ *@param {string} autocompleteUrl - The 
+ autoComplete Url.
+ *@param {object} options - Other options 
+ passed to initTagField
+ */
+function initTagField(id: number, autocompleteUrl: string, options: object) {
```
Comment (lb-) [inline client/src/includes/initTagField.ts:11]: I think id should be a string here - as it's used to get an element.

File: client/src/includes/initTagField.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        if (val && val[0] !== '"' && val.indexOf(' ') > -1) {
+          return '"' + val + '"';
+        }
+  
+        return val;
+      },
+      ...options,
+    };
+  
+    $('#' + id).tagit(finalOptions);
```
Comment (lb-) [inline client/src/includes/initTagField.ts:26]: Let's use a little less code if we can avoid it. 

Move the element selector to the top of the function and use non-jquery code to selector. 

Then we can return early and not do any logic we don't need. 
```
const tagFieldElement = document.getElementById(I'd);

if (!tagFieldElement) {
  return;
}
```

File: client/src/includes/initTagField.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  window.initTagField = initTagField;
```
Comment (lb-) [inline client/src/includes/initTagField.ts:29]: Setting the global is not needed here, it's set when it's included in the core js

File: client/src/includes/initTagField.ts (hunk 1 above)
Comment (xomcodes) [reply client/src/includes/initTagField.ts]: Done

**Comment by lb-** [comment]
@xomcodes are you ok to mark this as ready for review?

**Comment by xomcodes** [comment]
Not yet please. I'm still working on it.

File: client/src/includes/initTagField.ts (hunk 1 above)
Comment (xomcodes) [reply client/src/includes/initTagField.ts]: Okay. I've done that.

File: client/src/includes/initTagField.ts (hunk 2 above)
Comment (xomcodes) [reply client/src/includes/initTagField.ts]: Done

**Comment by xomcodes** [comment]
my test keeps failing @lb-

File: client/src/includes/initTagField.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  
+          return val;
+        },
+        ...options,
+      };
+    
+      // $('#' + id).tagit(finalOptions);
+    }
+
+    export { initTagField };
```
Comment (lb-) [inline client/src/includes/initTagField.ts:34]: For some reason, when I checked out your branch, this line was missing (the export) so not sure what happened.

Also, be sure to set your editor up correctly as it looks like your line feeds are incorrect. or at least run `npm run format` when finished editing and commit the output.

File: client/src/includes/initTagField.test.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { initTagField } from './initTagField';
+
+describe("initTagField", () => {
+    it("should return double quotes if a tag is not in quotes", () => {
+        expect(initTagField("abc", "https://github.com/wagtail/bakerydemo", {}),finalOptions).toBe('""')
+    })
+})
```
Comment (lb-) [inline client/src/includes/initTagField.test.js:7]: Not sure what happened but when I checked out your branch this file was missing.

Nonetheless, this test will not work as the `initTagField` does not actually return anything.

Instead what we want is to simply test that if we pass in a valid element ID the underlying jQuery tagit util gets called.

Here is a functional test suite, I know this is not the most helpful but it's really good to understand this and if you are happy with this you can apply/refine on a new commit on this branch.

* Firstly, we must import jquery and assign to the window.$ global, otherwise the test will always fail as it does not have access to `$`.
* Secondly, we need to set up some mocks so that `tagit` is a valid function, we cannot import this as it's a static JS file, but we do not really care about its implementation in this unit test, just that it is called. So we set up a mock function  that does nothing but allows us to check the element that was passed in and the params.
* We have two tests; firstly we check that the tagit library does not get called if we pass in an invalid element id
* The second test checks that tagit is called and what it is called with, if a valid element with the matching id exists in the DOM.

I hope this is helpful, sometimes mocks in Jest are a bit tricky to get working, but it's always good to start small with unit tests and build up to something more complex. Sometimes I write the test `it` statements before writing anything and think about what we want to test.

Also, the Jest docs are huge but they are invaluable here + a bit of stack overflow searching, here are some helpful links

* https://stackoverflow.com/questions/8834094/how-do-i-mock-a-jquery-plugin
* https://stackoverflow.com/questions/24984014/how-can-i-stop-property-does-not-exist-on-type-jquery-syntax-errors-when-using (for the jquery interface in the actual code file)
* https://jestjs.io/docs/jest-object


```javascript
import $ from 'jquery';

window.$ = $;

import { initTagField } from './initTagField';

describe('initTagField', () => {
  let element;

  const tagitMock = jest.fn(function tagitMockInner() {
    element = this;
  });

  window.$.fn.tagit = tagitMock;

  beforeEach(() => {
    element = null;
    jest.clearAllMocks();
  });

  it('should not call jQuery tagit if the element is not found', () => {
    expect(tagitMock).not.toHaveBeenCalled();

    initTagField('not-present');

    expect(tagitMock).not.toHaveBeenCalled();
  });

  it('should call jQuery tagit if the element is found', () => {
    expect(tagitMock).not.toHaveBeenCalled();

    document.body.innerHTML = `
    <main>
      <input id="tag-input" type="text" value="abc" />
    </main>
    `;

    initTagField('tag-input', '/path/to/autocomplete/', {
      someOther: 'option',
    });

    // check the jQuery instance is the correct element
    expect(element).toContain(document.getElementById('tag-input'));

    // check the tagit util was called correctly with supplied params
    expect(tagitMock).toHaveBeenCalledWith(
      expect.objectContaining({
        autocomplete: { source: '/path/to/autocomplete/' },
        someOther: 'option',
      }),
    );

    // check the supplied preprocessTag function
    const [{ preprocessTag }] = tagitMock.mock.calls[0];

    expect(preprocessTag).toBeInstanceOf(Function);

    expect(preprocessTag()).toEqual();
    expect(preprocessTag('"flat white"')).toEqual(`"flat white"`);
    expect(preprocessTag("'long black'")).toEqual(`"'long black'"`);
  });
});
```

**Comment by xomcodes** [comment]
@lb- Thank you so much for the feedback and links shared  
I've made all the changes and a re-write of the unit test as recommended. Everything works fine now.


unittest

**Comment by xomcodes** [comment]
ci/circleci: frontend keeps failing even after running `npm run format` and Prettier.
I get this error:
 16:60  error  Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead  @typescript-eslint/ban-types

I replace `{}` with `unknown`, `never` and `object` that still didn't solve the issue.

**Comment by xomcodes** [comment]
I've  resolved the failing issue with the ci/circleci:frontend following the first  step on this documentation https://smartdevpreneur.com/how-to-use-and-disable-typescript-eslint-ban-types/ (How to Disable @typescript-eslint/ban-types)

File: client/src/includes/initTagField.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+// import  $ from 'jquery'
+/** 
+ *This is a function initTagField
+ *@function
+ *@param {string} id - The object id.
+ *@param {string} autocompleteUrl - The 
+ autoComplete Url.
+ *@param {object} options - Other options 
+ passed to initTagField
+ */
```
[the remark this answers is not in the document]
Comment (xomcodes) [reply client/src/includes/initTagField.ts]: Thank you for this

File: client/src/includes/initTagField.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Initialises the tag fields using the jQuery tagit widget
+ *
+ *@param id - element id to initialise against
+ *@param source -  auto complete URL source
+ *@param options - Other options passed to jQuery tagit
+ */
+// eslint-disable-next-line @typescript-eslint/ban-types
+const initTagField = (id: string, source: string, options: {}): void => {
```
Comment (lb-) [inline client/src/includes/initTagField.ts:17]:
```suggestion
const initTagField = (id: string, source: string, options: Record<string, any>): void => {
```

Try this instead. This is a record (object) with any value.

File: client/src/includes/initTagField.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+/**
+ * Initialises the tag fields using the jQuery tagit widget
+ *
+ *@param id - element id to initialise against
+ *@param source -  auto complete URL source
+ *@param options - Other options passed to jQuery tagit
+ */
+// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (lb-) [inline client/src/includes/initTagField.ts:16]:
```suggestion
```

**Review by lb- — APPROVED** [review]
Working well - thanks for all the clean up and follow through @xomcodes  - I will get this rebased and merged in.

**Comment by xomcodes** [comment]
> Working well - thanks for all the clean up and follow through @xomcodes - I will get this rebased and merged in.

Thank you @lb- for helping me pull through. I want to learn more about unit testing. Are there any resources you could recommend?

**Comment by lb-** [comment]
@xomcodes 

https://martinfowler.com/articles/practical-test-pyramid.html - a must read to understand the types of tests and what they are for. 

https://medium.com/@dev.cprice/stop-testing-implementation-details-77a3528336af - good article, quite opinionated, but kind of explains how unit tests should focus on the outward API (now with the way we are doing jQuery here this is hard to follow to the letter). 

https://jestjs.io/docs/getting-started - Jest docs are solid and a must read as we use Jest in JStesting

Hopefully that's a good start.

**Comment by xomcodes** [comment]
> @xomcodes
> 
> https://martinfowler.com/articles/practical-test-pyramid.html - a must read to understand the types of tests and what they are for.
> 
> https://medium.com/@dev.cprice/stop-testing-implementation-details-77a3528336af - good article, quite opinionated, but kind of explains how unit tests should focus on the outward API (now with the way we are doing jQuery here this is hard to follow to the letter).
> 
> https://jestjs.io/docs/getting-started - Jest docs are solid and a must read as we use Jest in JStesting
> 
> Hopefully that's a good start.

Wow!!. Thank you for this @lb- 
These are great resources. Just what i wanted.  I will dive right into it
